### PR TITLE
refactor(engine): add optional wellLocation to all well access commands

### PR DESCRIPTION
**Pull request by mcous**
## Overview

In order to support LPC, a user needs to pick up and drop tips from a location other than the nominal tip location in order to verify that a given labware offset will work.

To support this usage, this PR adds an optional `wellLocation` field to all well-access commands, which allows an offset vector to be applied to a given well access request.

```js
{
  "data": {
    "commandType": "pickUpTip",
    "params": {
      "pipetteId": "abc",
      "labwareId": 123,
      "wellName": "A1",
      "wellLocation": {   // optional, default value {}
        "origin": "top",  // optional, default value "top"
        "offset": {       // optional, default value {}
          "x": 1,         // optional, default value 0
          "y": 2,         // optional, default value 0
          "z": 3,         // optional, default value 0
        }
      }
    }
  }
}

```

## Changelog

- refactor(engine): add optional wellLocation to all well access commands

The most interesting part of this change is that `dropTip` was already applying a `wellLocation` offset internally to move into the tiprack slightly to perform the drop in the correct location. To account for this, the drop tip will now ~sum any user-provided offset with its internal, z-only offset~ force the client-provided offset to reference `WellOrigin.TOP` and simply subtract the z-portion from the user-provided offset.

## Review requests

- [ ] Code looks good
- [ ] Tests look good
- [ ] Try this out if you can

## Risk assessment

Low, all things considered. Movement planning for relative well offsets was already in place for aspirate and dispense, so this was a pretty natural extension.

### CODE REVIEW

File: api/src/opentrons/protocol_engine/state/labware.py [hunk 1]
```diff
@@ -279,6 +282,37 @@ def get_well_rows(self, labware_id: str) -> Dict[str, List[str]]:
...
+
+        origin = result.origin
+        offset = result.offset
+        well_depth = self.get_well_definition(labware_id, well_name).depth
+
+        for loc in well_locations[1:]:
+            next_x = offset.x + loc.offset.x
+            next_y = offset.y + loc.offset.y
+
+            if origin == WellOrigin.TOP and loc.origin == WellOrigin.BOTTOM:
```
Comment (sanni-t) [inline api/src/opentrons/protocol_engine/state/labware.py]: Probably should note in the docstring that this 'addition' of locations is not commutative. The order of the locations passed will determine the result.

Comment (mcous) [reply api/src/opentrons/protocol_engine/state/labware.py]: After our chat, going to simplify this. General case summing of relative offsets is not what we need here

**Review by sanni-t — APPROVED** [review]
📐
Thanks for making the changes!

File: api/src/opentrons/protocol_engine/state/geometry.py [hunk 2]
```diff
@@ -173,21 +173,28 @@ def get_tip_geometry(
...
         """Get tip drop location given labware and hardware pipette."""
-        # return to top if labware is fixed trash
-        is_fixed_trash = self._labware.get_has_quirk(
-            labware_id=labware_id,
-            quirk="fixedTrash",
-        )
+        if well_location.origin != WellOrigin.TOP:
+            raise errors.WellOriginNotAllowedError(
+                'Drop tip location must be relative to "top"'
+            )
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/state/geometry.py:184]: 👍  Appreciate the domain checking here.

File: api/src/opentrons/protocol_engine/state/geometry.py [hunk 3]
```diff
@@ -173,21 +173,28 @@ def get_tip_geometry(
...
-        if is_fixed_trash:
-            return WellLocation()
+        # return to top if labware is fixed trash
+        if self._labware.get_has_quirk(labware_id=labware_id, quirk="fixedTrash"):
+            return well_location
 
         nominal_length = self._labware.get_tip_length(labware_id)
         offset_factor = pipette_config["return_tip_height"]
-        offset = WellOffset(x=0, y=0, z=-nominal_length * offset_factor)
+        tip_z_offset = nominal_length * offset_factor
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/state/geometry.py:192]: Is this intentionally multiplication, or should it be addition or subtraction?

Comment (mcous) [reply api/src/opentrons/protocol_engine/state/geometry.py:192]: Was and is intentionally multiplication

**Pull request merged**